pkg/k8s/identitybackend/identity.go

```diff
@@ -0,0 +1,433 @@
+// Copyright 2019 Authors of Cilium
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package identitybackend
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"reflect"
+	"strconv"
+	"strings"
+
+	"github.com/cilium/cilium/pkg/allocator"
+	"github.com/cilium/cilium/pkg/idpool"
+	"github.com/cilium/cilium/pkg/k8s/apis/cilium.io/v2"
+	clientset "github.com/cilium/cilium/pkg/k8s/client/clientset/versioned"
+	"github.com/cilium/cilium/pkg/k8s/informer"
+	"github.com/cilium/cilium/pkg/k8s/types"
+	k8sversion "github.com/cilium/cilium/pkg/k8s/version"
+	"github.com/cilium/cilium/pkg/kvstore"
+	"github.com/cilium/cilium/pkg/labels"
+	"github.com/cilium/cilium/pkg/logging"
+	"github.com/cilium/cilium/pkg/logging/logfields"
+	"github.com/sirupsen/logrus"
+
+	"k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/fields"
+	k8sTypes "k8s.io/apimachinery/pkg/types"
+	"k8s.io/client-go/tools/cache"
+)
+
+var (
+	log = logging.DefaultLogger.WithField(logfields.LogSubsys, "crd-allocator")
+)
+
+func NewCRDBackend(c CRDBackendConfiguration) (allocator.Backend, error) {
+	return &crdBackend{CRDBackendConfiguration: c}, nil
+}
+
+type CRDBackendConfiguration struct {
+	NodeName string
+	Store    cache.Store
+	Client   clientset.Interface
+	KeyType  allocator.AllocatorKey
+}
+
+type crdBackend struct {
+	CRDBackendConfiguration
+}
+
+func (c *crdBackend) DeleteAllKeys() {
+}
+
+// sanitizeK8sLabels strips the 'k8s:' prefix in the labels generated by
+// AllocatorKey.GetAsMap (when the key is k8s labels). In the CRD identity case
+// we map the labels directly to the ciliumidentity CRD instance, and
+// kubernetes does not allow ':' in the name of the label. These labels are not
+// the canonical labels of the identity, but used to ease interaction with the
+// CRD object.
+func sanitizeK8sLabels(old map[string]string) map[string]string {
+	k8sPrefix := labels.LabelSourceK8s + ":"
+	fixup := make(map[string]string, len(old))
+	for k, v := range old {
+		if !strings.HasPrefix(k, k8sPrefix) {
+			continue // skip non-k8s labels
```

Comment:
Perhaps log that we are skipping a non-k8s label? It would be useful to warn to users / signal to them somehow that we are skipping labels.

Reply:
What would you say here to avoid confusion? The meta labels are the ones that drop non-k8s labels. The SecurityLabels field would hold all. I'm not sure how to convey that a label being ignored does not mean that it isn't used in the identity.

Reply:
Hm, maybe say that these labels are skipped only when we are adding them to the ObjectMeta, which does not mean we are removing them from the labels (`SecurityLabels`) used in the actual identity itself.